sklearn/callback/tests/_utils.py

```diff
@@ -0,0 +1,107 @@
+# License: BSD 3 clause
+# Authors: the scikit-learn developers
+
+from sklearn.base import BaseEstimator, _fit_context, clone
+from sklearn.callback import BaseCallback
+from sklearn.callback._base import _eval_callbacks_on_fit_iter_end
+from sklearn.utils.parallel import Parallel, delayed
+
+
+class TestingCallback(BaseCallback):
+    def on_fit_begin(self, estimator, *, X=None, y=None):
+        pass
+
+    def on_fit_end(self):
+        pass
+
+    def on_fit_iter_end(self, estimator, node, **kwargs):
+        pass
+
+
+class TestingAutoPropagatedCallback(TestingCallback):
+    auto_propagate = True
+
+
+class NotValidCallback:
+    """Unvalid callback since it does not inherit from `BaseCallback`."""
+
+    def on_fit_begin(self, estimator, *, X=None, y=None):
+        pass
+
+    def on_fit_end(self):
+        pass
+
+    def on_fit_iter_end(self, estimator, node, **kwargs):
+        pass
+
+
+class Estimator(BaseEstimator):
+    _parameter_constraints: dict = {}
+
+    def __init__(self, max_iter=20):
+        self.max_iter = max_iter
+
+    @_fit_context(prefer_skip_nested_validation=False)
+    def fit(self, X, y):
+        root = self._eval_callbacks_on_fit_begin(
+            levels=[
+                {"descr": "fit", "max_iter": self.max_iter},
+                {"descr": "iter", "max_iter": None},
+            ],
+            X=X,
+            y=y,
+        )
+
+        for i in range(self.max_iter):
+            if _eval_callbacks_on_fit_iter_end(
+                estimator=self,
+                node=root.children[i],
+            ):
+                break
```

Comment:
Is there a way to cover this one?

Reply:
It will be covered with the EarlyStopping callback